gatsby-node.js

```diff
@@ -194,60 +84,179 @@ async function doQuery({
+    return {
+      ...groupedQueries,
+      [indexName]: [
+        ...(groupedQueries.hasOwnProperty(indexName)
+          ? groupedQueries[indexName]
+          : []),
+        queryOptions,
+      ],
+    };
+  }, {});
+}
 
-        if (!extObj) return true;
+/**
+ * Run all queries for a given index, then make any updates / removals necessary
+ */
+async function runIndexQueries(
+  indexName,
+  queries = [],
+  { client, activity, graphql, config }
+) {
+  const {
+    settings: mainSettings,
+    chunkSize = 1000,
+    enablePartialUpdates = false,
+    matchFields: mainMatchFields = ['modified'],
+  } = config;
 
-        return matchFields.some(field => extObj[field] !== curObj[field]);
-      });
+  setStatus(
+    activity,
+    `Running ${queries.length} ${
+      queries.length === 1 ? 'query' : 'queries'
+    } for index ${indexName}...`
+  );
 
-      Object.keys(algoliaObjects).forEach(objectID => {
-        // if the object has one of the matchFields, it should be removed,
-        // but objects without matchFields are considered "not controlled"
-        // and stay in the index
-        if (matchFields.some(field => algoliaObjects[objectID][field])) {
-          toRemove[objectID] = true;
-        }
-      });
-    }
+  const objectMapsByQuery = await Promise.all(
+    queries.map(query => getObjectsMapByQuery(query, graphql))
+  );
 
-    setQueryStatus(
-      `Partial updates – [insert/update: ${hasChanged.length}, total: ${objects.length}]`
+  const allObjectsMap = objectMapsByQuery.reduce((acc, objectsMap = {}) => {
+    return {
+      ...acc,
+      ...objectsMap,
+    };
+  }, {});
+
+  setStatus(
+    activity,
+    `${queries.length === 1 ? 'Query' : 'Queries'} resulted in a total of ${
+      Object.keys(allObjectsMap).length
+    } results`
+  );
+
+  const index = client.initIndex(indexName);
+  const tempIndex = client.initIndex(`${indexName}_tmp`);
+  const indexToUse = await getIndexToUse({
+    index,
+    tempIndex,
+    enablePartialUpdates,
+  });
+
+  let toIndex = {}; // used to track objects that should be added / updated
+  const toRemove = {}; // used to track objects that are stale and should be removed
+
+  // let objectsToIndex = [];
+
+  if (enablePartialUpdates !== true) {
+    // if enablePartialUpdates isn't true, so index all objects
+    toIndex = { ...allObjectsMap };
+    // objectsToIndex.push(...Object.values(allObjectsMap));
+  } else {
+    // iterate over each query to determine which data are fresh
+    setStatus(activity, `Starting Partial updates...`);
+
+    // get all match fields for all queries to minimize calls to the api
+    const allMatchFields = getAllMatchFields(queries, mainMatchFields);
+
+    // get all indexed objects matching all matched fields
+    const indexedObjects = await fetchAlgoliaObjects(
+      indexToUse,
+      allMatchFields
     );
+
+    // iterate over each query
+    for (const [i, { matchFields = mainMatchFields }] of queries.entries()) {
```

Comment:
it's a little confusing to see queries being used as the result of the queries, could it be called queryResults in all places maybe? WDYT?

Reply:
Yeah, I agree. I'm happy to rename